KAFKA-15178: Improve ConsumerCoordinator.poll perf

nicktelford opened the pull request:
When there are many topic-partitions being consumed by the application, the `rejoinNeededOrPending` method can perform poorly, due to the deep equality comparison of `metadataSnapshot` with `assignmentSnapshot`.

Since these snapshots can only diverge in `maybeUpdateSubscriptionMetadata`, we can use a `boolean` to cache whether they are already known to be "dirty", and therefore need to be compared.

This should shortcut the deep comparison in the common-case that the snapshot hasn't been updated.

Fixes KAFKA-15178

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

nicktelford committed:
KAFKA-15178: Improve ConsumerCoordinator.poll perf

When there are many topic-partitions being consumed by the application,
the `rejoinNeededOrPending` method can perform poorly, due to the deep
equality comparison of `metadataSnapshot` with `assignmentSnapshot`.

Since these snapshots can only diverge in `maybeUpdateSubscriptionMetadata`,
we can use a `boolean` to cache whether they are already known to be
"dirty", and therefore need to be compared.

This should shortcut the deep comparison in the common-case that the
snapshot hasn't been updated.

Fixes KAFKA-15178

ableegoldman commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -119,6 +119,7 @@ public final class ConsumerCoordinator extends AbstractCoordinator {
     private Set<String> joinedSubscription;
     private MetadataSnapshot metadataSnapshot;
     private MetadataSnapshot assignmentSnapshot;
+    private boolean metadataUpdated;
```
This optimization makes perfect sense but I am extremely paranoid that future changes will end up modifying the snapshot without remembering (or even knowing about) the need to update this boolean as well. Can you maybe encapsulate this in the `MetadataSnapshot` class or otherwise make sure it's impossible for this to get out of sync?

nicktelford replied:
Hmm, good point. I'm not sure how to encapsulate this inside `MetadataSnapshot` though, as the entire object is replaced when it's updated.

Would it be enough to place the swap in a private method, say:
```
private void updateAssignmentSnapshot() {
    assignmentSnapshot = metadataSnapshot;
    metadataUpdates = true;
}
```

nicktelford replied:
On reflection, I'm not convinced that this strategy will work, because there's already a shortcut on the `version` check, which means that the performance I'm seeing is when `assignmentSnapshot.version != metadataSnapshot.version`, so we definitely need to do the deep-equality check to determine if the new version of the metadata is actually different from the previous version.

I'm looking into some alternatives around whether we can represent this internal metadata differently, in a way that's more efficient to compare.

nicktelford replied:
I've revised the implementation to instead attempt to optimize the comparison itself in 63ee9ab, but I'm not entirely sure that it will perform better.

I'd like to construct a micro-benchmark for it, but I honestly have no idea how I'd run a micro-benchmark on this part of the codebase.

nicktelford replied:
Note: in 633f0e7 I switched to using a `LinkedHashSet`, because it performs better for iteration (which is used during `equals`) than `HashSet`, at the cost of a little more memory.

ableegoldman replied:
I think I would personally need to see benchmarks (somehow) to be convinced that just changing it to a `LinkedHashSet` is sufficient as an optimization, or even just guaranteed not to be a regression. Speaking from experience, I once tried to optimize the cache in Streams for better range-scan performance which involved swapping out the underlying data structure for a `LinkedHashMap`. The consequences were pretty dire, with the `LinkedHashMap` having comparably terrible performance characteristics when scaling up the keyspace. Granted it's not the exact same scenario because that was a map and also the Concurrent variant which surely means the scaling characteristics are not going to be exactly the same. But it has made me incredibly suspicious of the `LinkedHashXXX` data structures for use cases like this.

All that said, if this is literally only used for an equality check and nothing else, why not just hash it or something like that?

nicktelford replied:
That's interesting, although I suspect that may be more of a problem with `ConcurrentLinkedHash(Map|Set)` than `LinkedHash(Map|Set)`, because I would expect the doubly-linked list may require some locking or atomic heroics to make safe.

I did want to create a JMH micro-benchmark to test this, but I really struggled to do so because `MetadataSnapshot` and `PartitionRackInfo` are `private`, and micro-benchmarking the entire `ConsumerCoordinator` would require lots of mocking.

I might just remove the `private` modifier to enable the benchmark, temporarily, and not include it in these changes.

nicktelford committed:
Use flatter MetadataSnapshot structure

To optimize the performance of a deep-comparison of the `MetadataSnapshot`
structures, we re-organize it so that each topic corresponds to a `List`
of `partition`-`rack` tuples, instead of a nested `Set` of `racks` for
each `partition`.

This should reduce the impact of constructing a large number of
`HashSet` `Iterator`s, which appears to be where the bulk of the
performance problem is.

nicktelford committed:
Use a LinkedHashSet to improve performance

LinkedHashSet is faster for iteration than HashSet, as it just has to
follow some pointers.

Crucially, this makes the construction of the Iterators a constant-time
operation, which should remove most of the performance issues.

Memory usage is a bit higher, due to the doubly-linkedlist that's used
to maintain order.

showuon reviewed (commented):
@nicktelford , thanks for the PR! 
But I'm confused now. From the current change, I didn't see why this change improves the performance? I see it still need to go through O(partition_num) to check equality. Could you explain more about the current change? Thanks.

nicktelford commented:
> @nicktelford , thanks for the PR! But I'm confused now. From the current change, I didn't see why this change improves the performance? I see it still need to go through O(partition_num) to check equality. Could you explain more about the current change? Thanks.

@showuon My profiling shows that the biggest performance issue is constructing the `Iterator` for the `HashSet`s. Since, currently, there are two levels of `HashSet`, you need to construct `num_topic_partitions + topics` `Iterator` instances.

Flattening the structure ensures that we only construct 1 `Iterator`.

That said, I'm not _entirely_ convinced this will resolve the problem.

showuon commented:
@nicktelford , could you show the profiling result after this change, with the comparison in [KAFKA-15178](https://issues.apache.org/jira/browse/KAFKA-15178), so that we can assure this change will fix it.
